### PR TITLE
HPCC-20980 Automatically build and distribute class files for embedded Java code

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [ ] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Simple test case works, will add some others.

I think existing queries should be ok as the new method was added at the end.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Probably worth you starting to look

File: plugins/javaembed/javaembed.cpp [hunk 1]
```diff
@@ -3254,13 +3534,148 @@ extern DECL_EXPORT IEmbedContext* getEmbedContext()
...
 }
 
+static bool isValidIdentifier(const char *source)
+{
+    return isalnum(*source) || *source=='_' || *source=='$' || ::readUtf8Size(source)>1;   // This is not strictly accurate but probably good enough
+}
+
+static bool isFullClassFile(StringBuffer &className, size32_t len, const char *source)
+{
+    // A heuristic to determine whether the supplied embedded source is a full class file of just a single method
```
Comment (ghalliday) [inline plugins/javaembed/javaembed.cpp]: typo of->or

File: plugins/javaembed/javaembed.cpp [hunk 2]
```diff
@@ -3254,13 +3534,148 @@ extern DECL_EXPORT IEmbedContext* getEmbedContext()
...
+        source += csize;
+        len--;
+    }
+    // If we get here then it doesn't have a { at all - we COULD say it needs the prototype too but for now, who knows...
+    return false;
+}
+
+extern DECL_EXPORT void precompile(size32_t & __lenResult,void * & __result,size32_t charsBody,const char * body)
+{
+    unsigned lenBody = rtlUtf8Size(charsBody, body);  // size in bytes
```
Comment (ghalliday) [inline plugins/javaembed/javaembed.cpp]: trivial: possibly better named as sizeBody

File: plugins/javaembed/javaembed.cpp (hunk 1 above)
Comment (richardkchapman) [reply plugins/javaembed/javaembed.cpp]: Fixed

File: plugins/javaembed/javaembed.cpp (hunk 2 above)
Comment (richardkchapman) [reply plugins/javaembed/javaembed.cpp]: Renamed

File: plugins/javaembed/javaembed.cpp [hunk 3]
```diff
@@ -1597,11 +1601,11 @@ class JavaThreadContext
...
     }
 
-    void setThreadClassLoader(const char *classPath)
+    void setThreadClassLoader(const char *classPath, size32_t bytecodeLen, const byte *bytecode)
     {
-        if (classPath && *classPath)
+        if (bytecodeLen || (classPath && *classPath))
         {
-            if (prevClassPath && strcmp(classPath, prevClassPath) == 0)
+            if (!bytecodeLen && prevClassPath && classPath && strcmp(classPath, prevClassPath) == 0)  // MORE - caching of inline classes is important too...
```
Comment (ghalliday) [inline plugins/javaembed/javaembed.cpp]: Is there a jira to improve the caching?  Is it correct that precompiled is not currently cached?

**Review by ghalliday — COMMENTED** [review]
Nothing significant.  @richardkchapman please can you squash ready for merging

File: plugins/javaembed/javaembed.cpp [hunk 4]
```diff
@@ -130,6 +135,7 @@ static class JavaGlobalState
 
         // Options we know we always want set
         optionStrings.append("-Xrs");
+        //optionStrings.append("-XX:+TraceClassLoading");
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply plugins/javaembed/javaembed.cpp]: It's not really related to TRACE_CLASSFILE. I think OK to leave as is

File: plugins/javaembed/javaembed.cpp (hunk 3 above)
Comment (richardkchapman) [reply plugins/javaembed/javaembed.cpp]: This will be moot if I manage to implement the thread-local stuff. If I can't I will have to revisit.

**Pull request merged**